### ISSUE TITLE
--test-full enhancements and simplifications

### ISSUE THREAD

**Issue by jfoug** [magnumripper/JohnTheRipper#1792]
Several things.

1. self_tests and self_tests_full functions should be merged into a single function.  They are 99% the same, except for a few unique blocks of code which could easily be controlled with if logic.

2. change self_test_full to 'always' run with =0 for benchmarking, and use the optional number to allow different 'levels' of deep testing. Deeper tests with larger number will be newly 'found' cases to test, that may still have some formats which fail.  The -format-full=0 level should be 'expected' to not have any failures.

**Comment by jfoug** [comment]
@magnumripper,

I have sent of the email to Solar, and crickets are all that has happened.  Possibly you should bump it for me.   Is this code also going into core?

I really looked at the 2 functions, the self_test and self_test_full, and having 2 functions does NOT make any sense, other than if it will be that way in core, it will better match.  This really should be 1 function with a param, listing if it is full or not, and then some conditional logic (possibly move the special code out into external functions, so the amount of code options IN the main ST function is smaller.   But it really should not be 2 functions.   Also, I think using the optional number for different things between -test and -test-full makes a lot of sense.  Then -test 'really' becomes the benchmark method.  It also could be that an optional -benchmark param could be used, where -test 'defaults' to this being -benchmark=1 while -test-full simply defaults to -benchmark=0, but with an extra param, you 'could' either turn off bench-marking in -test, or turn it ON, in -test-full if you so chose to do it that way.

But I am still holding off on doing any coding for this, still hoping to hear 'something' from Solar (or you).

**Comment by magnumripper** [comment]
I'd really like Solar to merge the --test-full stuff (or declare that he never will) before going further. Doing what you suggest (indeed good suggestions) would make us deviate so much from core there would be no way back.

**Comment by jfoug** [comment]
Yes, that is WHY I have not moved forward.  It really 'should' be done, but unfortunately we have to be lead around with a ring in our nose.

**Comment by jfoug** [comment]
can you bump my email posted on the 28th, to try to spur a response.  Asking what you ask here 'might' be a good start.   I just always hate bumping my own emails.

**Comment by jfoug** [comment]
To be honest, the only ++-- merge would be the actual function declaration line (since we would add a param.  Other than that, I believe that 'all' changes would be +++++ blocks.

**Comment by magnumripper** [comment]
Let's give it a few days. Now we have this issue, we can relax and do other things for a while.

**Comment by jfoug** [comment]
@loverszhaokai since you are the one tasked with the original work for -test-full,  What are your thoughts here?

**Comment by magnumripper** [comment]
http://www.openwall.com/lists/john-dev/2015/09/30/9

@jfoug it's all yours to play with :smile:

**Comment by jfoug** [comment]
I figured if you poked him he might reply.

**Comment by jfoug** [comment]
It really should be 1 function.  I plan on this.

1.   there will be 3 options.   -test[=#]  -test-full[=#] -benchmark=#
2.  -test   is short for -test=1
3.  -test=x  is short for   -test=x  -benchmark=x
4.  -test-full is short for -test-full=0
5.  -test-full=#    Each # will run more tests (some of which may still have failures).
6.  -test-full=0 should 'assume' to run with no errors (unless someone adds/changes format to cause problems).
7.  --benchmark=xx 'could' be added to -test-full to actually make it benchmark (so you do 'not' lose that option).

The only change to logic, is that by default -test-full does not benchmark.  All other 'default' behavior is the same as it is today.  But this will 'gain' us ability to develop additional tests in -test-full in the 'hot' code, by simply putting that new logic into a section of the test function that is only run if the param to -test-full is > 0.  Then once we know all formats are 'fixed' or whitelisted for that behavior, we can move it into -test-full=0 codepath.

Does that sound like an acceptable plan?   It really should not be too hard to get this fit into John, AND I think it can be done with almost no 'collisions' in the fmt_test john-proper code base.

**Comment by magnumripper** [comment]
The only problem I see with this is we're very short of options flags. We bumped them to 64-bit but there are only a couple left IIRC. One day or the other we need to find out how to make them 128-bit yet portable :sunglasses:

**Comment by jfoug** [comment]
This should no need flags. I think there is enough there for the -test stuff.  Var's are all that is needed, to store the number.  We probably could just use the number for test (since they 'mean' the same thing).    I will have to dig in a little.  I just have a plan, I still need to really see how the code is laid out, and try to get it 'fit' into the current implementation as painless as possible.

**Comment by magnumripper** [comment]
The new --benchmark option definitely needs a flag

**Comment by jfoug** [comment]
We really should look at some other method than the 'flags' bits we use.  Their usefulness is over.  Time to use named enumeration strings and getter/setter functions.

**Comment by magnumripper** [comment]
The flags are primarily for not allowing "--wordlist --benchmark" and so on.

**Comment by jfoug** [comment]
Then instead of 

if (flags & SOME_BIT) { code; }

we would have

if (flags_bool_set("SOME_BIT")) {code;}

Same results, but we are not limited to 64 of them.

**Comment by jfoug** [comment]
But like I said, there are MUCH better ways to do that.

**Comment by magnumripper** [comment]
Come to think of it, I added logic for not needing a separate option flag for anything with OPT_REQ_PARAM. So you'd just use FLG_ZERO for the benchmark option and no new flag needed. The req_set for it would be `(FLG_TEST_CHK | FLG_TEST_FULL_CHK)`

**Comment by loverszhaokai** [comment]
@jfoug 

Thanks for your working with --test-full. :+1: 

My work has **not** been done with --test-full, since I have been preparing for interview and graduation in the following months. I can not spend much time on the work now, which I should have done during the GSOC. I hope to finish the work when I have time.

**The original motivation of introducing --test-full is**: 

1. Right now, --test performs only very basic testing, hashing one password at a time (albeit in different key indices).  It is not uncommon for a format being developed to pass self-test, yet fail to crack some passwords when tested for real on a password hash file and a wordlist.  We could have a longer-running, yet more extensive self-test and/or fuzzing invoked via e.g. --test-full or/and --fuzz options (to be introduced). 

2. The extensive test should mimic actual cracking (testing groups of mixed correct and incorrect passwords at once) and perhaps also **combine it with benchmarking**.  Right now, our --test starts with a quick self-test and then proceeds with a benchmark, which are separate stages; with an extensive **self-test that mimics actual cracking, the self-test and benchmark should be one and the same stage**.

>1. self_tests and self_tests_full functions should be merged into a single function. They are 99% the same, except for a few unique blocks of code which could easily be controlled with if logic.

Yes, self_test_full copies the logic of self_test. Most of them are the same. This is because I want to do a more extensive test based on the original self-test. I think we should merge the self_test_full into self_test when the self_test_full is finished and tested.

>2. change self_test_full to 'always' run with =0 for benchmarking, and use the optional number to allow different 'levels' of deep testing. Deeper tests with larger number will be newly 'found' cases to test, that may still have some formats which fail. The -format-full=0 level should be 'expected' to not have any failures.

Good ideas. "-format-full=0" should be "-test-full=0" ?

**Comment by jfoug** [comment]
@magnumripper I was actaully able to DROP the FLG_TEST_FULL_CHK flag totally, so it is now available. -test and -test-full are very similar. All flags for the full check are the same as for the normal check, EXCEPT I have to force the param (i..e it can not be optional), AND I have a 2nd variable to place it into. That variable is set to a default -1  This is due to the default of test being set to 1.   You will just have to see it when I get it in.

Now, I do have a few 'differences' from my original plan.

1. -test-full=req_param  (no longer optional)
2. -test-full is not currently allowing any benchmarking. That 'may' change, but for now, benchmark is set to 0

So instead of more flags used, you gain one, lol.

**Comment by magnumripper** [comment]
:+1:

**Comment by jfoug** [comment]
I did end up with 1 - merge issue in formats.c   The trailing 'while' part of the main do{}while() had to be made different.  -test and -test-full use completely different while blocks. SO I use a variable to 'hold' the condition, and simply while loop on that variable.

But in the end, that was the ONLY line that did not diff between bleeding and master in a clean manner.

Also, there was a flag for -test-full that I was able to eliminate.  What I did was make 2 int param variables for testing.  I kept the 'original' variable for -test.  Btw, it defaults to a 1 (1s). There is now a new param, for the deep test. I default it to -1. This made into a required param, so that we know if the var is not -1, then the user entered the -test-full=n command.

**Comment by jfoug** [comment]
There is certainly still additional work to be done for -test-full, but now it is sharing the function.

I 'started' a case-test4, which will be a 'generic' case bug finder.  I figure we scour the input string, looking for 'known length' strings that are hex and all of a single case (and not pure numeric). When we find them, we switch case, and then see if split fixes it back, OR if valid fails. If one of those cases happen, we are ok with that part of the hash being 'adjusted'.  If not, we fail with a split-case bug.     Now, that logic will likely have some false positives, so we may have to either change test case strings, OR white list a format or 2.

**Comment by jfoug** [comment]
I have a couple formats to 'fix' still for -test-full mode.  Once that is done, we need to see if we can figure out why this causes problems on the CI's.  It does not on my hardware.

But once test 4 is added, I think we have ALL of the formats now identified.

**Comment by jfoug** [comment]
This is now 100% incorporated, and all formats are updated (well all formats except ```@dynamic=``` format)  There are a couple of white lists.
